pyfar/signals/deterministic.py

```diff
@@ -242,6 +298,155 @@ def exponential_sweep(n_samples, frequency_range, n_fade_out=90, amplitude=1,
     return signal
 
 
+def exponential_sweep_freq(
+        n_samples, start_margin=None, stop_margin=None, frequency_range=None,
+        butterworth_order=8, double=True, sampling_rate=44100):
+
+    signal, group_delay = _sweep_synthesis_freq(
+        n_samples, "exponential", start_margin, stop_margin,
+        frequency_range, butterworth_order, double, sampling_rate)
+
+    return signal, group_delay
+
+
+def magnitude_weighted_sweep(
+        n_samples, magnitude, start_margin=None, stop_margin=None,
+        double=True, sampling_rate=44100):
+
+    signal, group_delay = _sweep_synthesis_freq(
+        n_samples, magnitude, start_margin, stop_margin,
+        None, None, double, sampling_rate)
+
+    return signal, group_delay
+
+
+def perfect_sweep(
+        n_samples, sampling_rate=44100):
+
+    signal, group_delay = _sweep_synthesis_freq(
+        n_samples, "perfect", None, None, None, None, False, sampling_rate)
+
+    return signal, group_delay
+
+
+def _sweep_synthesis_freq(
+        n_samples, magnitude, start_margin=None, stop_margin=None,
+        frequency_range=None, buterworth_order=8, double=True,
+        sampling_rate=44100):
+    """
+    Frequency domain sweep synthesis with arbitrary magnitude response.
+
+    TODO Link to fractional octave smoothing in Notes
+
+    TODO Implement calculation of group delay of impulse responses
+
+    TODO Examples
+
+    Sweep sweep synthesis according to [#]_
+
+    .. note::
+        The linear and exponential sweep can also be generated in the time
+        domain (see :py:func:`~linear_sweep`, :py:func:`~exponential_sweep`).
+        Frequency domain synthesis exhibits smooth magnitude spectra and in
+        trade of a slightly irregular temporal envelope. Time domain synthesis
+        exhibits a constant temporal envelope in trade of slight ripples in the
+        magnitude response.
+
+    Parameters
+    ----------
+    n_samples : int
+        The length of the sweep in samples.
+    magnitude : Signal, string
+        Specify the magnitude response of the sweep.
+
+        signal
+            The magnitude response as :py:class:`~pyfar.classes.audio.Signal`
+            object. If ``signal.n_samples`` is smaller than `n_samples`, zeros
+            are padded to the end of `signal`. Note that `frequency_range` is
+            not required in this case.
+        ``'linear'``
+            Design a sweep with linearly increasing frequency and a constant
+            magnitude spectrum.
+        ``'exponential'``
+            Design a sweep with exponentially increasing frequency. The
+            magnitude decreases by 3 dB per frequency doubling and has constant
+            energy in fiters of relative constant bandwidth (e.g. octaves).
+        ``'perfect'``
+            Perfect sweep according to [#]_. Note that the parameters
+            `start_margin`, `stop_margin`, `frequency_range` and `double` are
+            not required in this case.
```

Comment:
Calling this magnitude is a bit misleading. The effect on the magnitude is only a secondary effect.

Reply:
Renamed it to `sweep_type`. But it should not be critical any more since it was moved to a private function